File: backend/internal/handler/appv1/puzzle.go
package appv1

import (
	"context"

	"github.com/ishihaya/lateral-thinking-backend/domain/model"
	appv1 "github.com/ishihaya/lateral-thinking-backend/internal/handler/proto/app/v1"
)

func modelToProtoPuzzles(ps []*model.Puzzle) []*appv1.Puzzle {
	res := make([]*appv1.Puzzle, len(ps))
	for i, p := range ps {
		res[i] = &appv1.Puzzle{
			Id:       p.ID().String(),
			Question: p.Question(),
			Answer:   p.Answer(),
		}
	}
	return res
}

func (h *handler) ListPuzzles(ctx context.Context, req *appv1.ListPuzzlesRequest) (*appv1.ListPuzzlesResponse, error) {
	ps, err := h.puzzleUsecase.List(ctx, int(req.Limit))
	if err != nil {
		return nil, err
	}
	return &appv1.ListPuzzlesResponse{
		Puzzles: modelToProtoPuzzles(ps),
	}, nil
}

File: native/lib/infrastructure/repositoryimpl/protoconv/puzzle.dart
import 'package:lateral_thinking/domain/model/puzzle.dart';
import 'package:lateral_thinking/infrastructure/repositoryimpl/proto/app/v1/puzzle.pb.dart' as proto;

extension PuzzleProto on proto.Puzzle {
  Puzzle toModel() {
    return Puzzle(
      id: id,
      question: question,
      answer: answer,
    );
  }
}